// @testing-library/user-event/dist/cjs/utils/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var isClickableInput = require('./click/isClickableInput.js');
var Blob = require('./dataTransfer/Blob.js');
var DataTransfer = require('./dataTransfer/DataTransfer.js');
var FileList = require('./dataTransfer/FileList.js');
var Clipboard = require('./dataTransfer/Clipboard.js');
var getValue = require('./edit/getValue.js');
var input = require('./edit/input.js');
var isContentEditable = require('./edit/isContentEditable.js');
var isEditable = require('./edit/isEditable.js');
var setFiles = require('./edit/setFiles.js');
var walkRadio = require('./edit/walkRadio.js');
var blur = require('./focus/blur.js');
var copySelection = require('./focus/copySelection.js');
var cursor = require('./focus/cursor.js');
var focus = require('./focus/focus.js');
var getActiveElement = require('./focus/getActiveElement.js');
var getTabDestination = require('./focus/getTabDestination.js');
var isFocusable = require('./focus/isFocusable.js');
var selectAll = require('./focus/selectAll.js');
var resolveCaretPosition = require('./focus/resolveCaretPosition.js');
var selection = require('./focus/selection.js');
var selector = require('./focus/selector.js');
var readNextDescriptor = require('./keyDef/readNextDescriptor.js');
var cloneEvent = require('./misc/cloneEvent.js');
var eventWrapper = require('./misc/eventWrapper.js');
var findClosest = require('./misc/findClosest.js');
var getDocumentFromNode = require('./misc/getDocumentFromNode.js');
var getTreeDiff = require('./misc/getTreeDiff.js');
var getWindow = require('./misc/getWindow.js');
var isDescendantOrSelf = require('./misc/isDescendantOrSelf.js');
var isElementType = require('./misc/isElementType.js');
var isVisible = require('./misc/isVisible.js');
var isDisabled = require('./misc/isDisabled.js');
var level = require('./misc/level.js');
var wait = require('./misc/wait.js');
var cssPointerEvents = require('./pointer/cssPointerEvents.js');



exports.isClickableInput = isClickableInput.isClickableInput;
exports.readBlobText = Blob.readBlobText;
exports.createDataTransfer = DataTransfer.createDataTransfer;
exports.getBlobFromDataTransferItem = DataTransfer.getBlobFromDataTransferItem;
exports.createFileList = FileList.createFileList;
exports.attachClipboardStubToView = Clipboard.attachClipboardStubToView;
exports.createClipboardItem = Clipboard.createClipboardItem;
exports.detachClipboardStubFromView = Clipboard.detachClipboardStubFromView;
exports.readDataTransferFromClipboard = Clipboard.readDataTransferFromClipboard;
exports.resetClipboardStubOnView = Clipboard.resetClipboardStubOnView;
exports.writeDataTransferToClipboard = Clipboard.writeDataTransferToClipboard;
exports.getValue = getValue.getValue;
exports.input = input.input;
exports.getContentEditable = isContentEditable.getContentEditable;
exports.isContentEditable = isContentEditable.isContentEditable;
Object.defineProperty(exports, 'editableInputTypes', {
	enumerable: true,
	get: function () { return isEditable.editableInputTypes; }
});
exports.isEditable = isEditable.isEditable;
exports.isEditableInput = isEditable.isEditableInput;
exports.setFiles = setFiles.setFiles;
exports.walkRadio = walkRadio.walkRadio;
exports.blur = blur.blur;
exports.copySelection = copySelection.copySelection;
exports.getNextCursorPosition = cursor.getNextCursorPosition;
exports.focus = focus.focus;
exports.getActiveElement = getActiveElement.getActiveElement;
exports.getActiveElementOrBody = getActiveElement.getActiveElementOrBody;
exports.getTabDestination = getTabDestination.getTabDestination;
exports.isFocusable = isFocusable.isFocusable;
exports.isAllSelected = selectAll.isAllSelected;
exports.selectAll = selectAll.selectAll;
exports.resolveCaretPosition = resolveCaretPosition.resolveCaretPosition;
exports.getInputRange = selection.getInputRange;
exports.hasNoSelection = selection.hasNoSelection;
exports.hasOwnSelection = selection.hasOwnSelection;
exports.modifySelection = selection.modifySelection;
exports.modifySelectionPerMouseMove = selection.modifySelectionPerMouseMove;
exports.moveSelection = selection.moveSelection;
exports.setSelection = selection.setSelection;
exports.setSelectionPerMouseDown = selection.setSelectionPerMouseDown;
exports.setSelectionRange = selection.setSelectionRange;
exports.updateSelectionOnFocus = selection.updateSelectionOnFocus;
exports.FOCUSABLE_SELECTOR = selector.FOCUSABLE_SELECTOR;
exports.readNextDescriptor = readNextDescriptor.readNextDescriptor;
exports.cloneEvent = cloneEvent.cloneEvent;
exports.eventWrapper = eventWrapper.eventWrapper;
exports.findClosest = findClosest.findClosest;
exports.getDocumentFromNode = getDocumentFromNode.getDocumentFromNode;
exports.getTreeDiff = getTreeDiff.getTreeDiff;
exports.getWindow = getWindow.getWindow;
exports.isDescendantOrSelf = isDescendantOrSelf.isDescendantOrSelf;
exports.isElementType = isElementType.isElementType;
exports.isVisible = isVisible.isVisible;
exports.isDisabled = isDisabled.isDisabled;
Object.defineProperty(exports, 'ApiLevel', {
	enumerable: true,
	get: function () { return level.ApiLevel; }
});
exports.getLevelRef = level.getLevelRef;
exports.setLevelRef = level.setLevelRef;
exports.wait = wait.wait;
exports.assertPointerEvents = cssPointerEvents.assertPointerEvents;
exports.hasPointerEvents = cssPointerEvents.hasPointerEvents;
