// @testing-library/user-event/dist/cjs/options.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var keyMap = require('./keyboard/keyMap.js');
var keyMap$1 = require('./pointer/keyMap.js');

exports.PointerEventsCheckLevel = void 0;
(function(PointerEventsCheckLevel) {
    PointerEventsCheckLevel[PointerEventsCheckLevel[/**
   * Check pointer events on every user interaction that triggers a bunch of events.
   * E.g. once for releasing a mouse button even though this triggers `pointerup`, `mouseup`, `click`, etc...
   */ "EachTrigger"] = 4] = "EachTrigger";
    PointerEventsCheckLevel[PointerEventsCheckLevel[/** Check each target once per call to pointer (related) API */ "EachApiCall"] = 2] = "EachApiCall";
    PointerEventsCheckLevel[PointerEventsCheckLevel[/** Check each event target once */ "EachTarget"] = 1] = "EachTarget";
    PointerEventsCheckLevel[PointerEventsCheckLevel[/** No pointer events check */ "Never"] = 0] = "Never";
})(exports.PointerEventsCheckLevel || (exports.PointerEventsCheckLevel = {}));
/**
 * Default options applied when API is called per `userEvent.anyApi()`
 */ const defaultOptionsDirect = {
    applyAccept: true,
    autoModify: true,
    delay: 0,
    document: globalThis.document,
    keyboardMap: keyMap.defaultKeyMap,
    pointerMap: keyMap$1.defaultKeyMap,
    pointerEventsCheck: exports.PointerEventsCheckLevel.EachApiCall,
    skipAutoClose: false,
    skipClick: false,
    skipHover: false,
    writeToClipboard: false,
    advanceTimers: ()=>Promise.resolve()
};
/**
 * Default options applied when API is called per `userEvent().anyApi()`
 */ const defaultOptionsSetup = {
    ...defaultOptionsDirect,
    writeToClipboard: true
};

exports.defaultOptionsDirect = defaultOptionsDirect;
exports.defaultOptionsSetup = defaultOptionsSetup;
