// @testing-library/user-event/dist/esm/utils/index.js
export { isClickableInput } from './click/isClickableInput.js';
export { readBlobText } from './dataTransfer/Blob.js';
export { createDataTransfer, getBlobFromDataTransferItem } from './dataTransfer/DataTransfer.js';
export { createFileList } from './dataTransfer/FileList.js';
export { attachClipboardStubToView, createClipboardItem, detachClipboardStubFromView, readDataTransferFromClipboard, resetClipboardStubOnView, writeDataTransferToClipboard } from './dataTransfer/Clipboard.js';
export { getValue } from './edit/getValue.js';
export { input } from './edit/input.js';
export { getContentEditable, isContentEditable } from './edit/isContentEditable.js';
export { editableInputTypes, isEditable, isEditableInput } from './edit/isEditable.js';
export { setFiles } from './edit/setFiles.js';
export { walkRadio } from './edit/walkRadio.js';
export { blur } from './focus/blur.js';
export { copySelection } from './focus/copySelection.js';
export { getNextCursorPosition } from './focus/cursor.js';
export { focus } from './focus/focus.js';
export { getActiveElement, getActiveElementOrBody } from './focus/getActiveElement.js';
export { getTabDestination } from './focus/getTabDestination.js';
export { isFocusable } from './focus/isFocusable.js';
export { isAllSelected, selectAll } from './focus/selectAll.js';
export { resolveCaretPosition } from './focus/resolveCaretPosition.js';
export { getInputRange, hasNoSelection, hasOwnSelection, modifySelection, modifySelectionPerMouseMove, moveSelection, setSelection, setSelectionPerMouseDown, setSelectionRange, updateSelectionOnFocus } from './focus/selection.js';
export { FOCUSABLE_SELECTOR } from './focus/selector.js';
export { readNextDescriptor } from './keyDef/readNextDescriptor.js';
export { cloneEvent } from './misc/cloneEvent.js';
export { eventWrapper } from './misc/eventWrapper.js';
export { findClosest } from './misc/findClosest.js';
export { getDocumentFromNode } from './misc/getDocumentFromNode.js';
export { getTreeDiff } from './misc/getTreeDiff.js';
export { getWindow } from './misc/getWindow.js';
export { isDescendantOrSelf } from './misc/isDescendantOrSelf.js';
export { isElementType } from './misc/isElementType.js';
export { isVisible } from './misc/isVisible.js';
export { isDisabled } from './misc/isDisabled.js';
export { ApiLevel, getLevelRef, setLevelRef } from './misc/level.js';
export { wait } from './misc/wait.js';
export { assertPointerEvents, hasPointerEvents } from './pointer/cssPointerEvents.js';
