// @testing-library/user-event/dist/cjs/utils/edit/isEditable.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var isElementType = require('../misc/isElementType.js');
var isContentEditable = require('./isContentEditable.js');

function isEditable(element) {
    return isEditableInput(element) || isElementType.isElementType(element, 'textarea', {
        readOnly: false
    }) || isContentEditable.isContentEditable(element);
}
exports.editableInputTypes = void 0;
(function(editableInputTypes) {
    editableInputTypes['text'] = 'text';
    editableInputTypes['date'] = 'date';
    editableInputTypes['datetime-local'] = 'datetime-local';
    editableInputTypes['email'] = 'email';
    editableInputTypes['month'] = 'month';
    editableInputTypes['number'] = 'number';
    editableInputTypes['password'] = 'password';
    editableInputTypes['search'] = 'search';
    editableInputTypes['tel'] = 'tel';
    editableInputTypes['time'] = 'time';
    editableInputTypes['url'] = 'url';
    editableInputTypes['week'] = 'week';
})(exports.editableInputTypes || (exports.editableInputTypes = {}));
function isEditableInput(element) {
    return isElementType.isElementType(element, 'input', {
        readOnly: false
    }) && Boolean(exports.editableInputTypes[element.type]);
}

exports.isEditable = isEditable;
exports.isEditableInput = isEditableInput;
